flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java

```diff
@@ -0,0 +1,439 @@
+		 */
+		public void startResourceManager() throws Exception {
+			resourceManager.start();
+			rmServices.grantLeadership();
+		}
+
+		/**
+		 * Register a job master with the RM.
+		 */
+		public void registerJobMaster(MockJobMaster jobMaster) throws Exception  {
+			CompletableFuture<RegistrationResponse> registration = resourceManager.registerJobManager(
+					jobMaster.jobMasterId, jobMaster.resourceID, jobMaster.address, jobMaster.jobID, timeout);
+			assertTrue(registration.get() instanceof JobMasterRegistrationSuccess);
+		}
+
+		@Override
+		public void close() throws Exception {
+			rpcService.stopService();
+		}
+	}
+
+	static class TestContainer extends UtilsTest.TestingContainer {
+		Resource resource;
+		Priority priority;
+
+		TestContainer(String id, String host) {
+			super(id, host);
+		}
+
+		@Override
+		public Resource getResource() {
+			return resource;
+		}
+
+		@Override
+		public void setResource(Resource resource) {
+			this.resource = resource;
+		}
+
+		@Override
+		public Priority getPriority() {
+			return priority;
+		}
+
+		@Override
+		public void setPriority(Priority priority) {
+			this.priority = priority;
+		}
+	}
+
+	@Test
+	public void testAllocateNewContainer() throws Exception {
+		new Context() {{
+			startResourceManager();
+			resourceManager.startNewWorker(resourceProfile1);
+			verify(mockResourceManagerClient).addContainerRequest(any(AMRMClient.ContainerRequest.class));
+			Container testingContainer = new TestContainer(task1.toString(), "host1");
+			testingContainer.setResource(Resource.newInstance(200, 1));
+			testingContainer.setPriority(Priority.UNDEFINED);
+			resourceManager.onContainersAllocated(ImmutableList.of(testingContainer));
+			assertThat(resourceManager.workerNodeMap,
+					Matchers.hasKey(testingContainer.getId().toString()));
+			verify(mockNMClient).startContainer(eq(testingContainer), any(ContainerLaunchContext.class));
+		}};
+	}
+
+	@Test
+	public void testStopWorker() throws Exception {
+		new Context() {{
+			startResourceManager();
+			resourceManager.startNewWorker(resourceProfile1);
+			verify(mockResourceManagerClient).addContainerRequest(any(AMRMClient.ContainerRequest.class));
+			Container testingContainer = new TestContainer(task1.toString(), "host1");
+			testingContainer.setResource(Resource.newInstance(200, 1));
+			testingContainer.setPriority(Priority.UNDEFINED);
+			resourceManager.onContainersAllocated(ImmutableList.of(testingContainer));
+			assertThat(resourceManager.workerNodeMap,
+					Matchers.hasKey(testingContainer.getId().toString()));
+			verify(mockNMClient).startContainer(eq(testingContainer), any(ContainerLaunchContext.class));
+			resourceManager.stopWorker(new YarnWorkerNode(testingContainer));
+			verify(mockResourceManagerClient).releaseAssignedContainer(any(ContainerId.class));
+			assertThat(resourceManager.workerNodeMap,
+					not(Matchers.hasKey(testingContainer.getId().toString())));
+		}};
+	}
```

Comment:
You've spent a lot of effort implementing this test. I'm wondering whether we could make it a bit simpler because we do a lot of setup for testing that the method call `ResourceManager#stopWorker` calls `AMRMClientAsync#releaseAssignedContainer`. What we should test is that we actually return containers in the Yarn case but this is a bit tricky. In order to properly test this we would have to submit a Yarn application which starts a `YarnResourceManager` with a special testing `SlotManager`, for example. We could use the `SlotManager` to issue container allocations and container release calls. Additionally, we would have to check whether these containers are really released or not from within the application master.